pkg/kubelet/dockertools/docker.go

```diff
@@ -180,11 +180,30 @@ func matchImageTagOrSHA(inspected dockertypes.ImageInspect, image string) bool {
 	}
 
 	if isDigested {
-		algo := digest.Digest().Algorithm().String()
-		sha := digest.Digest().Hex()
-		// Look specifically for short and long sha(s)
-		if strings.Contains(inspected.ID, algo+":"+sha) {
-			// We found the short or long SHA requested
+		for _, repoDigest := range inspected.RepoDigests {
+			named, err := dockerref.ParseNamed(repoDigest)
+			if err != nil {
+				glog.V(4).Infof("couldn't parse image RepoDigest reference %q: %v", repoDigest, err)
+				continue
+			}
+			if d, isDigested := named.(dockerref.Digested); isDigested {
+				// Allow prefix matches for shorter SHA values
+				if digest.Digest().Algorithm().String() == d.Digest().Algorithm().String() &&
+					strings.HasPrefix(digest.Digest().Hex(), d.Digest().Hex()) {
+					return true
+				}
+			}
+		}
+
+		// process the ID as a digest
+		id, err := dockerdigest.ParseDigest(inspected.ID)
+		if err != nil {
+			glog.V(4).Infof("couldn't parse image ID reference %q: %v", id, err)
+			return false
+		}
+		// Allow prefix matches for shorter SHA values
+		if digest.Digest().Algorithm().String() == id.Algorithm().String() &&
+			strings.HasPrefix(digest.Digest().Hex(), id.Hex()) {
```

Comment:
I think we should not support prefixes.  I understand the request, but
Docker rejects it anyway.  Is there a concern that is too strict?

On Wed, Sep 14, 2016 at 11:56 AM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/kubelet/dockertools/docker.go
> <https://github.com/kubernetes/kubernetes/pull/32627#discussion_r78778411>
> :
>
> > +				if digest.Digest().Algorithm().String() == d.Digest().Algorithm().String() &&
> > +					strings.HasPrefix(digest.Digest().Hex(), d.Digest().Hex()) {
> > +					return true
> > +				}
> > +			}
> > +		}
> > +
> > +		// process the ID as a digest
> > +		id, err := dockerdigest.ParseDigest(inspected.ID)
> > +		if err != nil {
> > +			glog.V(4).Infof("couldn't parse image ID reference %q: %v", id, err)
> > +			return false
> > +		}
> > +		// Allow prefix matches for shorter SHA values
> > +		if digest.Digest().Algorithm().String() == id.Algorithm().String() &&
> > +			strings.HasPrefix(digest.Digest().Hex(), id.Hex()) {
>
> Same comment as above about prefixes
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32627/files/0c9e99486c388106a97f6c3491e15e54de53b97f#r78778411>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3VMY_gdg9QMHrHlwMumylrswyikks5qqBlHgaJpZM4J8SYk>
> .
>